core/src/collections/DynamicValue.hpp

```diff
@@ -28,47 +28,130 @@
-            std::shared_ptr<DynamicObject> object;
 
-            api::DynamicType type;
+        struct DynamicValue {
+            // Mandatory default constructor for it to be default-constructible in (e.g.) maps.
+            explicit DynamicValue();
+            explicit DynamicValue(const DynamicValue& rhs);
+            explicit DynamicValue(const char* x);
+            explicit DynamicValue(const std::string& x);
+            explicit DynamicValue(const std::vector<uint8_t>& x);
+            explicit DynamicValue(bool x);
+            explicit DynamicValue(int32_t x);
+            explicit DynamicValue(int64_t x);
+            explicit DynamicValue(double x);
+            explicit DynamicValue(const std::shared_ptr<DynamicArray>& x);
+            explicit DynamicValue(const std::shared_ptr<DynamicObject>& x);
 
             template <class Archive>
             void serialize(Archive& ar) {
-
             }
 
             std::string dump() const;
             std::ostream& dump(std::ostream& ss, int depth) const;
-        };
 
-        template<> void DynamicValue::serialize<cereal::PortableBinaryOutputArchive>( cereal::PortableBinaryOutputArchive & );
-        template<> void DynamicValue::serialize<cereal::PortableBinaryInputArchive>( cereal::PortableBinaryInputArchive & );
-        template<> void DynamicValue::serialize<cereal::BinaryOutputArchive>( cereal::BinaryOutputArchive & );
-        template<> void DynamicValue::serialize<cereal::BinaryInputArchive>( cereal::BinaryInputArchive & );
+            /// Get the underlying type.
+            api::DynamicType getType() const;
+
+            /// Change the value to a std::string by copying.
+            DynamicValue& operator=(const std::string& rhs);
+            /// Change the value to a std::string by moving in.
+            DynamicValue& operator=(std::string&& rhs);
+
+            /// Change the value to a std::vector<uint8_t> by copying.
+            DynamicValue& operator=(const std::vector<uint8_t>& rhs);
+            /// Change the value to a std::vector<uint8_t> by moving in.
+            DynamicValue& operator=(std::vector<uint8_t>&& rhs);
+
+            /// Change the value to a bool by copying.
+            DynamicValue& operator=(bool rhs);
+
+            /// Change the value to a int32_t by copying.
+            DynamicValue& operator=(int32_t rhs);
+
+            /// Change the value to a int64_t by copying.
+            DynamicValue& operator=(int64_t rhs);
+
+            /// Change the value to a double by copying.
+            DynamicValue& operator=(double rhs);
+
+            /// Change the value to a DynamicArray by copying.
+            DynamicValue& operator=(const std::shared_ptr<DynamicArray>& rhs);
 
+            /// Change the value to a DynamicObject by copying.
+            DynamicValue& operator=(const std::shared_ptr<DynamicObject>& rhs);
+
+            /// Try to get the value as if it were a std::string.
+            optional<std::string&> asStr();
```

Comment:
```
template<typename T>
optional<T> as() {
   if (data.type() == typeid(T)) return boos::get<T>(data);
   return optional<T>();
}
```

Reply:
Look at https://www.boost.org/doc/libs/1_65_0/doc/html/boost_typeindex.html

Reply:
I know that and I know RTTI is big and heavy and I wish we could get rid of it. However, there is a way to go with it without RTTI and still removing that `api::DynamicType` stuff directly — using the type index in the variant, which is a bit silly since you might have issues if you had types in the middle of the variant.